src/connection-loop.ts

```diff
@@ -0,0 +1,229 @@
+
+const MAX_CONNECTIONS = 3;
+
+type Duration = {duration: number; ok: boolean};
+type Durations = Duration[];
+
+export interface ConnectionLoopDelegate {
+  closed(): boolean;
+  invokeExecute(): unknown;
+  debounceDelay?: number;
+  maxConnections?: number;
+  printTime(): string;
+  watchdogTimer?: number;
+}
+
+export class ConnectionLoop {
+  private _pendingResolver = resolver<void>();
+  private _currentResolver = resolver<void>();
+  private _pushCounter = 0;
+
+  private readonly _delegate: ConnectionLoopDelegate;
+  private _lastSendTime = 0;
+
+  constructor(delegate: ConnectionLoopDelegate) {
+    this._delegate = delegate;
+    this.run();
+  }
+
+  async execute(): Promise<void> {
+    this._pendingResolver.resolve();
+    // await a promise to allow the currentPushResolver to be created inside the
+    // connection loop.
+    await Promise.resolve(0);
+    await this._currentResolver.promise;
+  }
+
+  async run(): Promise<void> {
+    const durations: Durations = [];
+    let delay = MIN_DELAY;
+    let lastConnectionErrored = false;
+    const delegate = this._delegate;
+    const {
+      debounceDelay = DEBOUNCE_DELAY,
+      maxConnections = MAX_CONNECTIONS,
+      watchdogTimer,
+    } = delegate;
+
+    console.log('Starting connection loop', delegate.printTime());
+
+    while (!delegate.closed()) {
+      console.log('Waiting for a push', delegate.printTime());
+
+      if (watchdogTimer) {
+        await Promise.race([
+          this._pendingResolver.promise,
+          sleep(watchdogTimer),
+        ]);
+      } else {
+        await this._pendingResolver.promise;
+      }
+
+      // This resolvers is used to make the individual calls to push have the
+      // correct "duration".
+      const currentResolver = resolver();
+      this._currentResolver = currentResolver;
+      // Don't let this rejected promise escape.
+      currentResolver.promise.catch(() => 0);
+
+      if (debounceDelay) {
+        await sleep(debounceDelay);
+      }
+
+      // This resolver is used to wait for incoming push calls.
+      this._pendingResolver = resolver();
+
+      if (lastConnectionErrored || this._pushCounter >= maxConnections) {
+        console.log(
+          'Too many pushes. Waiting until one finishes',
+          delegate.printTime(),
+        );
+        await this._waitUntilAvailableConnection();
+      }
+
+      if (lastConnectionErrored || this._pushCounter > 0) {
+        if (lastConnectionErrored) {
+          console.log(
+            'Last connection errored. Sleeping for',
+            delay,
+            'ms',
+            delegate.printTime(),
+          );
+        } else {
+          console.log(
+            'More than one outstanding connection (',
+            this._pushCounter,
+            '). Sleeping for',
+            delay,
+            'ms',
+            delegate.printTime(),
+          );
+        }
+
+        if (lastConnectionErrored) {
+          await sleep(delay);
```

Comment:
This logic does not seem correct to me. Why do we wait the full `delay` in the case of the last connection errored, but do the subtraction otherwise? Seems they should be the same. The last connection having errored in my mind shouldn't change anything (except for how delay is computed).